### ISSUE TITLE
Doomsday Fuel Python Test 8

### ISSUE THREAD

**Issue by Hazerdous** [ivanseed/google-foobar-help#8]
Like others before, I'm having some trouble with doomsday_fuel. I've done my research on Markov Chains and feel like I have a pretty good understanding of them, and I'm using that to solve the problem.

I feel like I have everything right, but I'm missing test 8. I'm covering 1x1 chains, there don't seem to be any 2x2s (there are only 2 possible solutions to a 2x2, and I've tried hard-coding both)

Any ideas what I could be getting wrong?

**Comment by ivanseed** [comment]
Hey, have you tried to attempt to find the size of the matrix that is causing the test to fail?

stevem99 managed to determine the exact inputs of the test, and it turned out that his error was in determining the LCD when simplifying the answer.

**Comment by Hazerdous** [comment]
I've determined the size of the tests are as follows:
1 - len(5)
2,3 - len(6)
4 - len(1)
5,6,7,9,10 - len(10)
8 Can't be len(1), there is only 1 possible output in that situation, which I am handling.
So it could be 5, 6, or 10, and I am properly handling other test cases with those sizes.
I have also determined that no test has more than 6 terminal states, so there is no 10-terminals edge case.

I feel like the LCD section of my code is the least error-prone, unless something else is generating an invalid fraction, but if that were the case I believe all/most other cases would also fail.
I am generating a `range(denominator, 2**31-1, denominator)`, looping through the one with the lowest starting value, and testing if all ranges include that number. If so, that is the LCD.

**Comment by Dazby** [comment]
Why not use Euclid's method for finding the LCD? The method itself is for two numbers but you can find a way to expand that easy enough. It's guaranteed to return the correct LCD.

Also, if it's length 1, are you returning the function argument over 1?

Wish I was able to get back on the challenge again 😫

**Comment by marshalma** [comment]
It really helps to know the size of test case 4 is 1. I've got them all, luckily.

**Comment by ivanseed** [comment]
🤔  Do they not shuffle the test cases around though?

**Comment by marshalma** [comment]
It seems like they don't..🤣

**Comment by Marcus-L** [comment]
@ivanseed how did @stevem99 find the exact test inputs? @Hazerdous the hint in ivanseed's summary about ordering the states should help with number 8.

**Comment by ivanseed** [comment]
@Marcus-L 

> You will want to throw some conditionals against the argument that you are taking from each test. So for example you are being passed a map m.
> 
> If you write `if len(m) > 4` and `return null` inside the conditional you will fail all tests which the input was greater than 4 in length. Now lets say you find out by the process of elimination that all your tests pass except number 6 when you write `if len(m) == 1` - you still do not know with 100% certainty that the conditional is branching into null.
> 
> However you can replace the `return null` with `print('foo')` which will cause ANY tests that HITS this line to throw an error. If you get an error, you have found some information about the input that is causing you trouble. If you do not get an error, then that input is not being used by any tests.

**Comment by zhrts** [comment]
Can anyone point me towards a reliable matrix inversion method for python or java? I have found none...

**Comment by marshalma** [comment]
@zhrts I guess you are using Fractions so numpy won't work... I found one that works well at least on the testcases... http://stackoverflow.com/questions/32114054/matrix-inversion-without-numpy

**Comment by TickyMick** [comment]
Ok. Seems like for Test 4 the size is 1 and for test 6 there are no terminal states. I fail both. What should you return if it's size 1? 'cause I return an empty array and it fails. :(

**Comment by marshalma** [comment]
@TickyMick I hard programmed [1, 1] for test 4 and it passed... I guess it makes sense.

**Comment by TickyMick** [comment]
Wow. Yes it makes sense. :D Now I feel like an idiot.  What about the case where there are no terminal states?

**Comment by zhrts** [comment]
I can't get a single matrix inversion function working for fractions ;-; I store them as two-length java, arrays, is this bad? Fifteen hours left, I'll trade a neat mapping function that I never needed for a working matrix inverter...

**Comment by TickyMick** [comment]
Does anyone know if it is anything specific with test 6?

**Comment by Hazerdous** [comment]
@zhrts I used the `fractions.Fraction` class, all of the operations work directly, so `Fraction(1,2)/2` returns `Fraction(1,4)`, etc.

As for inversion, I recreated all of the mathematical steps for row operations, you can probably find a good tutorial through Google that explains the process step-by-step.

Of course, I had problems with one test so *something* I did was wrong, so I'd encourage finding a different method incase either of those were my issue.

**Comment by TickyMick** [comment]
Test 6 - any ideas? *stressed*

**Comment by zhrts** [comment]
@Hazerdous what do you make of this? 0_o
https://github.com/Trismeg/FooBar/blob/master/doomsday_fuel.py

**Comment by TickyMick** [comment]
@marshalma, @ivanseed Test 6 - any ideas? (๑•﹏•)⋆* ⁑⋆*
Test 9, that has same 10 states from which 6 are terminal, passes, but test 6 doesn't.
What could be wrong if all the tests pass except this one?
15 h left...:(

**Comment by TickyMick** [comment]
My method of finding the common denominator was for 2 fractions only, not for a list of fractions this is why it failed the 6th test.

**Comment by stevem99** [comment]
http://pastebin.com/HAiEwnD2

**Comment by Robuko** [comment]
In test case 6, the lowest common multiple of the denominators is not the largest denominator.  I believe that in all other cases, it is.

**Comment by j9ac9k** [comment]
I struggled with test case 8 on this challenge for a while, the unique thing about test case 8 is that it's the one challenge that demands that you swap rows/columns in the correct order, where in the other challenges, you can half-ass that bit and all will be well.

@Marcus-L made a comment further up the thread that points to the same thing.

TL;DR swap rows/cols correctly to get that test to pass.

**Comment by RGrun** [comment]
I too am having trouble with calculating the matrix inversions. @Hazerdous do you remember which mathematical method you used? I'm trying to code one out in Java and it seems like the Gauss-Jordan method would be the most straightforward.

**Comment by pickfire** [comment]
I tried another method which is using the path traversal method (traverse each non-terminal path), I eventually pass test 1-4 and test 7, I guess that method might work as well but just don't require much knowledge on the algorithm. Anyway, after reading the tips, I only know that it is using absorbing markov chain and solved it in one pass. Thanks a lot.

It took me like hours do understand what is the question and need some googing to know that it is markov chain by googling the program content.

**Comment by sasanghavi** [comment]
Sorry to dig up the thread on the Doomsday Fuel question. 
I am working with Java and am failing at test case 5,6 and 10.
If anyone has encountered this before, could you point me in the direction of something that I might have overlooked in the problem.

**Comment by patlub** [comment]
These look like all the test cases
https://sskaje.me/2017/05/googles-foo-bar-doomsday-fuel/

**Comment by sasanghavi** [comment]
@patlub  These helped. Apparently, the order of states I got after sorting was messed up from my end  for the non-terminal states. Somehow, test cases 5,6 and 10 were the ones which could catch the error. Thanks!

**Comment by dmasood** [comment]
Does anyone have a good matrix inversion method for Java? I am using the one here: http://www.sanfoundry.com/java-program-find-inverse-matrix/ and it does the job, but on the larger test cases I get time limit exceeded

**Comment by j9ac9k** [comment]
So the trick here is you don't actually want to invert the matrix.  Solving
for x in Ax = b can be done in several ways (one of which is inverting the
matrix).  You can try solving for x using Gaussian elimination (what I did)
or using crammers rule.  Both of those methods should require significantly
less code that doing matrix inversion.

Remember that when doing scientific computing, you almost never want to
invert a matrix; it just provides a nice method when writing things out,
but it's computationally expensive, there is no guarantee you even can
invert a matrix, and even if you do, there are cases that's hard to
discover where the solution is very unstable).

**Comment by dmasood** [comment]
@j9ac9k Can I do Gaussian elimination/Cramer's rule on the matrix that I currently have before the inversion step (from following the steps on ivanseeds readme i.e. the standard form)? I don't really understand the Maths behind the methods you mentioned just yet, so I will have to do some research. Thanks

**Comment by j9ac9k** [comment]
I guess I should just leave it with a pro-tip of forget trying to invert
the (I-Q) matrix, build the matrix up, and then append a column of F to the
end, and then perform Gaussian elimination.  You should end up with a bunch
of 0's and one non-zero value, where the non-zero value corresponds to the
steady state probability.

**Comment by jmelo11** [comment]
im stuck at matrix 8, im using fractions and they seem to explode, what the logic for swaping the rows and columns= @j9ac9k
